src/provider/systemdmanager/service.rs

```diff
@@ -0,0 +1,321 @@
+// unit files, I'll leave it in for now
+#[allow(dead_code)]
+static SECTION_ORDER: phf::OrderedSet<&'static str> =
+    phf_ordered_set! {"Unit", "Service", "Install"};
+
+pub struct Service {
+    pub systemd_units: Vec<SystemDUnit>,
+}
+
+impl Service {
+    pub fn new(pod: &Pod, pod_state: &PodState) -> Result<Self, StackableError> {
+        // Create systemd unit template with values we need from the pod object
+        let pod_settings = SystemDUnit::new_from_pod(&pod);
+
+        // Convert all containers to systemd unit files
+        let systemd_units: Result<Vec<SystemDUnit>, StackableError> = pod
+            .containers()
+            .iter()
+            .map(|container| SystemDUnit::new(&pod_settings, container, pod_state))
+            .collect();
+
+        systemd_units.map(|units| Self {
+            systemd_units: units,
+        })
+    }
+}
+
+#[derive(Clone)]
+pub struct SystemDUnit {
+    name: String,
+    sections: HashMap<String, HashMap<String, String>>,
+    environment: HashMap<String, String>,
+}
+
+impl SystemDUnit {
+    /// Create a new unit which inherits all common elements from ['common_properties'] and parses
+    /// everything else from the ['container']
+
+    pub fn new(
+        common_properties: &SystemDUnit,
+        container: &Container,
+        pod_state: &PodState,
+    ) -> Result<Self, StackableError> {
+        let mut unit = common_properties.clone();
+        unit.name = String::from(container.name());
+
+        unit.add_property("Unit", "Description", &unit.name.clone());
+
+        unit.add_property(
+            "Service",
+            "ExecStart",
+            &SystemDUnit::get_command(container, pod_state)?,
+        );
+
+        let env_vars = SystemDUnit::get_environment(container, pod_state)?;
+
+        for (name, value) in env_vars {
+            unit.add_env_var(&name, &value);
+        }
+
+        unit.add_property("Service", "StandardOutput", "journal");
+        unit.add_property("Service", "StandardError", "journal");
+        unit.add_property("Install", "WantedBy", "multi-user.target");
+
+        Ok(unit)
+    }
+
+    pub fn new_from_pod(pod: &Pod) -> Self {
+        let mut unit = SystemDUnit {
+            name: pod.name().to_string(),
+            sections: Default::default(),
+            environment: Default::default(),
+        };
+
+        let restart_policy = match &pod.as_kube_pod().spec {
+            Some(spec) => spec.restart_policy.as_deref().unwrap_or("Never"),
+            None => "Never",
+        };
+
+        unit.add_property(
+            "Service",
+            "Restart",
+            RESTART_POLICY_MAP.get(restart_policy).unwrap(),
+        );
+        unit
+    }
+
+    // Add a key=value entry to the specified section
+    fn add_property(&mut self, section: &'static str, key: &str, value: &str) {
+        let section = self
+            .sections
+            .entry(String::from(section))
+            .or_insert_with(HashMap::new);
```

Comment:
I actually prefer this, because your variant would require the reader to know or look up what type of an value _sections_  needs - which is defined somewhere else entirely. Not a large difference, but might help a little.